### PR TITLE
[fix] deployment scripts

### PR DESCRIPTION
**Pull request by josojo**
The current deployment scripts have a bug: During the deployment for non-testnet networks, the scripts would try to work with the local artifacts, rather than the artifacts from the imported npm packages with the correct addresses.

This PR fixes this by introducing the functions `getDeployedDependency` and `getArtificat`. Both functions first try to get the dependency from the local build folder and fall back to the imports, if the first method was not successful.

---

testplan:
run migration against local testnetwork
run migration against rinkeby
run migration of https://github.com/gnosis/dex-contracts-integration-example
=>
```
cd dex-contracts
yarn install
yarn link
git clone https://github.com/gnosis/dex-contracts-integration-example
cd dex-contracts-integration-example
yarn install
npx truffle migrate --reset
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,7 +66,7 @@
   "devDependencies": {
     "@gnosis.pm/mock-contract": "^3.0.8",
     "@gnosis.pm/owl-token": "^3.1.0",
-    "@gnosis.pm/solidity-data-structures": "^1.2.2",
+    "@gnosis.pm/solidity-data-structures": "^1.2.4",
```
Comment (fleupold) [inline package.json]: unrelated change 😞

File: src/migration/PoC_dfusion.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ async function migrate({ artifacts, deployer, network, account, web3, maxTokens
       web3,
     })
   }
-  const TokenOWLProxy = getDependency(
+  const TokenOWLProxy = await getDependency(
```
Comment (fleupold) [inline src/migration/PoC_dfusion.js]: It's weird that those variable names are upper case.

File: src/migration/utilities.js [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
-  // If this migration script is used from the repository dex-contracts, the contract
-  // data is received via the artificats.require.
-  // If this migration script is used from an external project, the first try statement
-  // will fail and it will get the contracts from the function initializeContract.
+  // The following logic ensures the right artifacts are used, no matter whether the migration scripts are run from an
+  // external project or this dex-contracts project.
   try {
     contract = artifacts.require(path.split("/").pop())
+    if (contractNeedsToBeDeployed) {
+      await contract.deployed()
```
Comment (fleupold) [inline src/migration/utilities.js]: Is this really deploying it? Or getting the deployed instance? I find it hacky that we introduce this `contractNeedsToBeDeployed` flag just for this one case. Can we not call deployed directly on the call site?

**Review by fleupold — COMMENTED** [review]
Why do we not want this for the BatchExchange? What's the effect of this to local migrations (not from the external project)?

File: src/migration/utilities.js (hunk 3 above)
Comment (josojo) [reply src/migration/utilities.js]: This line is not supposed to do any deployments or stuff like that. 
It really is just checking, whether the contract is already deployed in the current network.

But yeah, it is super confusing, though it works.
I will have to rethink all of it.

**Comment by josojo** [comment]
With my local truffle@5.0.24, it all works.

But I agree that the current method is too complex, I will have to rethink the whole migration strategies...

File: src/migration/utilities.js (hunk 3 above)
Comment (josojo) [reply src/migration/utilities.js]: Now, we are really getting a deployed dependency. I hope it is clearer now.

**Comment by josojo** [comment]
Finally, I fixed all the different bugs.

I needed to use the following linking method in tests:
`    await BatchExchange.link("IdToAddressBiMap", lib1.address)
`
instead of
`    await BatchExchange.link(IdToAddressBiMap, lib1.address)
`
as otherwise, I would get:
 Error: Cannot link the contract without an address. Cf. https://travis-ci.org/gnosis/dex-contracts/builds/642488383?utm_source=github_status&utm_medium=notification


Also, I am required to add 
`  await deployer.link(SnappBaseCore, SnappBase)
`
into the migration scripts, as otherwise the events would not be logged correctly.  Cf.
https://travis-ci.org/gnosis/dex-contracts/builds/642533397?utm_source=github_status&utm_medium=notification
If I find time this week, I will report this as a truffle issue.

it took forever to hunt this down. It would be great if we could merge this as intermediate fix and then I would try to hunt down the real reasons later, unless one of you guys has a clue about what is going and or has a better suggestion.

File: src/migration/utilities.js [hunk 4]
```diff
@@ -1,23 +1,33 @@
...
   return contract
 }
 
-function getDependency(artifacts, network, deployer, account, path) {
+async function getDeployedDependency(artifacts, deployer, account, path) {
+  let contract
+  // The following logic tries to get the deployed dependency from the local build folder, and falls back to
+  // npm imports. This makes sure that the script can be used within dex-contracts and from external projects.
+  try {
+    contract = artifacts.require(path.split("/").pop())
```
Comment (fleupold) [inline src/migration/utilities.js:20]: I'd expect this to be Contract and then `let contract = await Contract.deployed()`. Why are we running deployed but not assigning the result to the variable we return?

> MyContract.deployed()
Creates an instance of the contract abstraction representing the contract at its deployed address. The deployed address is a special value given to truffle-contract that, when set, saves the address internally so that the deployed address can be inferred from the given Ethereum network being used. This allows you to write code referring to a specific deployed contract without having to manage those addresses yourself. Like at(), deployed() is thenable, and will resolve to a contract abstraction instance representing the deployed contract after ensuring that code exists at that location and that that address exists on the network being used.

(https://github.com/trufflesuite/trufflesuite.com/blob/master/src/docs/truffle/reference/contract-abstractions.md)

File: src/migration/utilities.js [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
+    await contract.deployed()
+  } catch (error) {
+    contract = initializeContract(path, deployer, account)
+    await contract.deployed()
+  }
+  return contract
+}
+
+function getArtifact(artifacts, deployer, account, path) {
```
Comment (fleupold) [inline src/migration/utilities.js:29]: nit: contract should be called artifact in this method. Also `initializeContract` should be `initializeArtifact`

File: src/migration/utilities.js [hunk 6]
```diff
@@ -1,23 +1,33 @@
...
-    from: accounts[0],
+    from: account,
     gas: 6.5e6,
   })
   return contract
 }
 
-function getDependency(artifacts, network, deployer, account, path) {
+async function getDeployedDependency(artifacts, deployer, account, path) {
+  let contract
```
Comment (fleupold) [inline src/migration/utilities.js:16]: Can this method reuse `getArtifact`?

File: test/snapp/snapp_base.js [hunk 7]
```diff
@@ -30,11 +30,11 @@ contract("SnappBase", async accounts => {
   beforeEach(async () => {
     const lib1 = await IdToAddressBiMap.new()
 
-    await SnappBaseCore.link(IdToAddressBiMap, lib1.address)
+    await SnappBaseCore.link("IdToAddressBiMap", lib1.address)
```
Comment (fleupold) [inline test/snapp/snapp_base.js:33]: Why all these changes?

File: src/migration/utilities.js [hunk 8]
```diff
@@ -1,23 +1,33 @@
-function initializeContract(path, deployer, accounts) {
-  const Contract = require("truffle-contract")
+function initializeContract(path, deployer, account) {
+  const Contract = require("@truffle/contract")
```
Comment (fleupold) [inline src/migration/utilities.js:2]: ubernit: could be on top of file like normal imports.

File: src/migration/PoC_dfusion.js [hunk 9]
```diff
@@ -11,48 +11,34 @@ async function migrate({ artifacts, deployer, network, account, web3, maxTokens
       web3,
     })
   }
-  const TokenOWLProxy = getDependency(
+  const feeToken = await getDeployedDependency(
     artifacts,
-    network,
```
Comment (fleupold) [inline src/migration/PoC_dfusion.js:16]: Nice, loving to see this go

File: src/migration/PoC_dfusion.js [hunk 10]
```diff
@@ -11,48 +11,34 @@ async function migrate({ artifacts, deployer, network, account, web3, maxTokens
...
   )
 
-  // Hack to populate truffle artifact data correctly for linked libraries.
-  await BiMap.deployed()
-  await IterableAppendOnlySet.deployed()
-
   //linking libraries
-  await deployer.link(BiMap, BatchExchange)
-  await deployer.link(IterableAppendOnlySet, BatchExchange)
+  await deployer.link(biMap, BatchExchange)
```
Comment (fleupold) [inline src/migration/PoC_dfusion.js:36]: Does `BatchExchange.link(biMap)` do the same thing?

File: src/migration/PoC_dfusion.js [hunk 11]
```diff
@@ -11,48 +11,34 @@ async function migrate({ artifacts, deployer, network, account, web3, maxTokens
...
+  await deployer.link(iterableAppendOnlySet, BatchExchange)
 
   // eslint-disable-next-line no-console
   console.log("Deploy BatchExchange contract")
-  await deployer.deploy(BatchExchange, maxTokens, fee_token.address)
+  await deployer.deploy(BatchExchange, maxTokens, feeToken.address)
```
Comment (fleupold) [inline src/migration/PoC_dfusion.js:41]: So an external project would still deploy a new instance of BatchExchange? How would they be able to use the already deployed one?

**Comment by fleupold** [comment]
When trying to run the testplan I'm getting:

> Error: Error: Error: Cannot find module '@truffle/contract'

on `npx truffle migrate` inside dex-contracts-integration-example

**Comment by fleupold** [comment]
The testplan should include running `yarn install` before running `yarn link`. This way the error I mentioned above goes away.

~Now I'm running into:~

```
2_dev_dependencies.js
=====================
Error: Error: TypeError: Cannot read property '0' of undefined
    at Object.run (/Users/felixleupold/Gnosis/dex-contracts-integration-example/node_modules/truffle/build/webpack:/packages/truffle-migrate/index.js:92:1)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Truffle v5.0.24 (core: 5.0.24)
Node v10.18.0
```

**Update** This one was my fault. I was based off a bad branch.

File: src/migration/utilities.js (hunk 4 above)
Comment (josojo) [reply src/migration/utilities.js:20]: >Why are we running deployed but not assigning the result to the variable we return?

To my understanding, no assignment is needed, as `.deployed` just writes the address of the deployed contract into the current artifact.

File: src/migration/utilities.js (hunk 6 above)
Comment (josojo) [reply src/migration/utilities.js:16]: I was not able to do it, as here, we don't need `await contract.deployed()` in the try and catch

File: test/snapp/snapp_base.js (hunk 7 above)
Comment (josojo) [reply test/snapp/snapp_base.js:33]: I tried to explain it in the comment:

>I needed to use the following linking method in tests:
await BatchExchange.link("IdToAddressBiMap", lib1.address)
instead of
await BatchExchange.link(IdToAddressBiMap, lib1.address)
as otherwise, I would get:
Error: Cannot link the contract without an address. Cf. https://travis-ci.org/gnosis/dex-contracts/builds/642488383?utm_source=github_status&utm_medium=notification

**Comment by bh2smith** [comment]
The instructions provided for the test plan seem to imply that `dex-contracts-integration-example` is somehow contained in `dex-contracts`. I have simply attempted to do `cd ../dex-contracts-integration-example`, but then ran into a failure on `yarn install` which was 

```
error Package "" refers to a non-existing file '"/Users/alexherrmann/gnosis/dex-contracts"'.
```

(suggesting there is a local file reference from your computer).

**Comment by fleupold** [comment]
@bh2smith replace the local path with v0.2.0

you also need to change this line: https://github.com/gnosis/dex-contracts-integration-example/blob/cc57275752beb085640823a533919c653238bd10/migrations/2_dev_dependencies.js#L8

to `account: accounts[0]`

Same here: https://github.com/gnosis/dex-contracts-integration-example/blob/cc57275752beb085640823a533919c653238bd10/migrations/3_batch_exchange.js#L8

I managed to run the deployment script on ganache now. Moving on to rinkeby...

**Review by fleupold — CHANGES_REQUESTED** [review]
Looks like running the `dex-contracts-integration` migration on rinkeby leads to a newly deployed instance (https://rinkeby.etherscan.io/address/0x008ECeaBc0F7D26388024946f9B03b03219423B9). This is however likely to be a bug in the integration project and not related to this PR. How would one use the existing dƒusion instance? 

Also, I'm confused that we need to run to migrations `2_dev_dependencies.js` and `3_batch_exchange.js`. I'd expect projects that depend on dfusion to just call one migrateBatchExchange and that would take care of everything. Can we expose a single migration script that takes care of dependencies and the exchange itself? Why do we have two separate migration scripts in the first place?

File: src/migration/PoC_dfusion.js (hunk 10 above)
Comment (josojo) [reply src/migration/PoC_dfusion.js:36]: During migrations, truffle expects you to do the linking with deployer, and for tests you would use your command.

But I am not sure about the inner workings of why it is like that

File: test/snapp/snapp_base.js (hunk 7 above)
Comment (fleupold) [reply test/snapp/snapp_base.js:33]: But why? What difference does it make? I'd like to understand these changes better.

File: src/migration/utilities.js (hunk 4 above)
Comment (fleupold) [reply src/migration/utilities.js:20]: Here is the code: https://github.com/trufflesuite/truffle/blob/develop/packages/contract/lib/contract/constructorMethods.js#L87

You are right that it mutates this (which is awkward), so you don't need the assigned value. I'd still argue it's cleaner to use the assigned value in case this changes in the future.
